metadata/dashboard/dashboard_metadata.go

```diff
@@ -1074,16 +1121,219 @@ func (m *MetadataServer) getSourceDataIterator(name, variant string, limit int64
+		m.SetFoundVariantJSON(foundVariant, err, c, resourceType)
+	case "training-sets":
+		foundVariant, err := m.client.GetTrainingSetVariant(context.Background(), nameVariant)
+		m.SetFoundVariantJSON(foundVariant, err, c, resourceType)
+	case "sources":
+		foundVariant, err := m.client.GetSourceVariant(context.Background(), nameVariant)
+		m.SetFoundVariantJSON(foundVariant, err, c, resourceType)
+	case "entities":
+		foundVariant, err := m.client.GetEntity(context.Background(), name)
+		m.SetFoundVariantJSON(foundVariant, err, c, resourceType)
+	case "users":
+		foundVariant, err := m.client.GetUser(context.Background(), name)
+		m.SetFoundVariantJSON(foundVariant, err, c, resourceType)
+	case "models":
+		foundVariant, err := m.client.GetModel(context.Background(), name)
+		m.SetFoundVariantJSON(foundVariant, err, c, resourceType)
+	case "providers":
+		foundVariant, err := m.client.GetProvider(context.Background(), name)
+		m.SetFoundVariantJSON(foundVariant, err, c, resourceType)
+	}
+}
+
+type TagPostBody struct {
+	Tags    []string `json:"tags"`
+	Variant string   `json:"variant"`
+}
+
+func getResourceType(resourceTypeString string) metadata.ResourceType {
+	var resourceType metadata.ResourceType
+	switch resourceTypeString {
+	case "features":
+		resourceType = metadata.FEATURE_VARIANT
+	case "labels":
+		resourceType = metadata.LABEL_VARIANT
+	case "training-sets":
+		resourceType = metadata.TRAINING_SET_VARIANT
+	case "sources":
+		resourceType = metadata.SOURCE_VARIANT
+	case "entities":
+		resourceType = metadata.ENTITY
+	case "users":
+		resourceType = metadata.USER
+	case "models":
+		resourceType = metadata.MODEL
+	case "providers":
+		resourceType = metadata.PROVIDER
+	}
+	return resourceType
+}
+
+func (m *MetadataServer) PostTags(c *gin.Context) {
+	var requestBody TagPostBody
+	resourceTypeParam := c.Param("type")
+	if err := c.BindJSON(&requestBody); err != nil {
+		fetchError := m.GetTagError(500, err, c, "PostTags - Error binding the request body")
+		c.JSON(fetchError.StatusCode, fetchError.Error())
+		return
+	}
+	resourceType := getResourceType(resourceTypeParam)
+	name := c.Param("resource")
+	variant := requestBody.Variant
+
+	objID := metadata.ResourceID{
+		Name:    name,
+		Variant: variant,
+		Type:    resourceType,
+	}
+	foundResource, err := m.lookup.Lookup(objID)
+
+	if err != nil {
+		fetchError := m.GetTagError(400, err, c, "PostTags - Error finding the resource with resourceID")
+		c.JSON(fetchError.StatusCode, fetchError.Error())
+		return
+	}
+
+	replaceTags(resourceTypeParam, foundResource, &pb.Tags{Tag: requestBody.Tags})
+
+	m.lookup.Set(objID, foundResource)
+
+	c.JSON(http.StatusOK, TagResult{
+		Name:    name,
+		Variant: variant,
+		Tags:    requestBody.Tags,
+	})
+}
+
+func replaceTags(resourceTypeParam string, currentResource metadata.Resource, newTagList *pb.Tags) error {
+	deserialized := currentResource.Proto()
+	switch resourceTypeParam {
```

Comment:
Feel like this could be condensed with an interface. Ok for now but may want to rework if theres time

Reply:
100% yeah looking for a clean way to handle this. The issue is that interfaces in go only define _behavior_ and not properties. In this case I'd need common access to `*variant.Tags` which is a property in all the variant structs, so a quick interface won't work. 

Alternatively...I could write a method on all the variants, `setTags([]string)` and then create a single method interface, which gets rid of the switch but adds more boilerplate to metadata.go :(

Open to any ideas!